### PR TITLE
SONARPY-103 Migrate text colorizing to use Highlightable API

### PR DESCRIPTION
**Pull request by ivandalbosco**

### CODE REVIEW

File: sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonHighlighter.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+import org.sonar.api.batch.sensor.highlighting.NewHighlighting;
+import org.sonar.api.batch.sensor.highlighting.TypeOfText;
+import org.sonar.python.api.PythonKeyword;
+import org.sonar.python.api.PythonTokenType;
+import org.sonar.squidbridge.SquidAstVisitor;
+
+/**
+ * Colorizes Python code. Currently colorizes:
+ * <ul>
+ *   <li>String literals. Example: 34
```
Comment (vilchik-elena) [inline sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonHighlighter.java]: @ivandalbosco 34 is not best string literal :)

File: sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonHighlighter.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+ *   </li>
+ *   <li>Keywords. Example: def
+ *   <li>Comments. Example: # some comment
```
Comment (vilchik-elena) [inline sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonHighlighter.java]: @ivandalbosco imho it's better to provide code samples inside some tag, e.g. `<pre>` or `<code>` if one line

File: sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonHighlighter.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+ *   </li>
+ *   <li>Doc strings. Examples: """ a doc string""", ''' another doc string '''
```
Comment (vilchik-elena) [inline sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonHighlighter.java]: @ivandalbosco minor thing, but imho either we should format xml like this
```
<tag>content</tag>
```
or
```
<tag>
  content
</tag>
```

File: sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonHighlighter.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+  private final SensorContext context;
+  
+  public PythonHighlighter(SensorContext context) {
+    this.context = context;
+  }
+
+  @Override
+  public void visitFile(@Nullable AstNode astNode) {
+    newHighlighting = context.newHighlighting();
+    System.out.println("high = " + newHighlighting);
```
Comment (vilchik-elena) [inline sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonHighlighter.java]: @ivandalbosco do not forget to remove

File: sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonHighlighter.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    System.out.println("fs = " + context.fileSystem());
+    InputFile inputFile = context.fileSystem().inputFile(context.fileSystem().predicates().is(getContext().getFile().getAbsoluteFile()));
+    newHighlighting.onFile(inputFile);
+  }
+
+  @Override
+  public void visitToken(Token token) {
+    if (token.getType().equals(PythonTokenType.STRING)) {
+      // case: string literal
```
Comment (vilchik-elena) [inline sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonHighlighter.java]: @ivandalbosco imho these comments inside each branch are excessive

File: sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonHighlighter.java (hunk 1 above)
Comment (ivandalbosco) [inline sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonHighlighter.java]: Agreed.

File: sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonHighlighter.java (hunk 2 above)
Comment (ivandalbosco) [inline sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonHighlighter.java]: OK

File: sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonHighlighter.java (hunk 4 above)
Comment (ivandalbosco) [inline sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonHighlighter.java]: OK

File: sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonHighlighter.java (hunk 5 above)
Comment (ivandalbosco) [inline sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonHighlighter.java]: I'd rather keep them. When looking at the code, the reader goers straights away to the point

Comment (ivandalbosco) [inline sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonHighlighter.java]: OK Elena. Changes applied and pushed

File: sonar-python-plugin/src/test/java/org/sonar/plugins/python/PythonHighlighterTest.java [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+  }
+  
+  @Test
+  public void comment_misplaced() throws Exception {
+    List<TypeOfText> typeOfTexts = getTypesOfText(9, 18, 1);
+    
+    assertThat(typeOfTexts.get(0)).isEqualTo(TypeOfText.COMMENT);
+  }
+  
+  private List<TypeOfText>  getTypesOfText(int line, int column, int expectedSize) {
```
Comment (vilchik-elena) [inline sonar-python-plugin/src/test/java/org/sonar/plugins/python/PythonHighlighterTest.java]: @ivandalbosco we can drop last parameter

File: sonar-python-plugin/src/test/java/org/sonar/plugins/python/PythonHighlighterTest.java [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+
+  @Test
+  public void keyword_def() throws Exception {
+    List<TypeOfText> typeOfTexts = getTypesOfText(1, 1, 1);
+
+    assertThat(typeOfTexts.get(0)).isEqualTo(TypeOfText.KEYWORD);
+  }
+
+  @Test
+  public void keyword_four() throws Exception {
```
Comment (vilchik-elena) [inline sonar-python-plugin/src/test/java/org/sonar/plugins/python/PythonHighlighterTest.java]: @ivandalbosco i don't understand the name of this method

File: sonar-python-plugin/src/test/resources/org/sonar/plugins/python/highlighting.py [hunk 8]
```diff
@@ -0,0 +1,13 @@
+def foo():
+    pass
+
+x = "string"
+
+# This is a comment
+
+def foo():
+    pass  # misplaced comment
```
Comment (vilchik-elena) [inline sonar-python-plugin/src/test/resources/org/sonar/plugins/python/highlighting.py]: @ivandalbosco why this comment is misplaced?

File: sonar-python-plugin/src/test/java/org/sonar/plugins/python/PythonHighlighterTest.java [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+import org.sonar.api.batch.sensor.internal.SensorContextTester;
+import org.sonar.python.PythonAstScanner;
+
+import static org.fest.assertions.Assertions.assertThat;
+
+public class PythonHighlighterTest {
+  
+  private SensorContextTester context;
+  
+  private File file;
```
Comment (vilchik-elena) [inline sonar-python-plugin/src/test/java/org/sonar/plugins/python/PythonHighlighterTest.java]: @ivandalbosco can we make it local var?

File: sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonHighlighter.java [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+      highlight(token, TypeOfText.STRING);
+    } else if (token.getType() instanceof PythonKeyword) {
+      // case: keyword
+      highlight(token, TypeOfText.KEYWORD);
+    } else if (token.getType().equals(GenericTokenType.COMMENT)) {
+      // case: doc string
+      highlight(token, TypeOfText.COMMENT);
+    } 
+    
+    if (token.hasTrivia()) {
```
Comment (vilchik-elena) [inline sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonHighlighter.java]: @ivandalbosco You can drop this line as `token.getTrivia()` can return empty list

File: sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonHighlighter.java [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+      for (Trivia trivia : token.getTrivia()) {
```
Comment (vilchik-elena) [inline sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonHighlighter.java]: @ivandalbosco let's extract this code to separate method, e.g. `highlightComments`

File: sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonHighlighter.java [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+        if (trivia.isComment()) {
```
Comment (vilchik-elena) [inline sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonHighlighter.java]: @ivandalbosco we don't need this check. every trivia is comment

File: sonar-python-plugin/src/test/java/org/sonar/plugins/python/PythonHighlighterTest.java [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+  private File file; 
+  
+  @Before
+  @SuppressWarnings("unchecked")
+  public void scanFile() {
+    file = new File("src/test/resources/org/sonar/plugins/python/highlighting.py");
+    DefaultInputFile inputFile = new DefaultInputFile("moduleKey", file.getName())
+      .initMetadata(new FileMetadata().readMetadata(file, Charsets.UTF_8));
+    
+    context = SensorContextTester.create(new File("src/test/resources/org/sonar/plugins/python").getAbsoluteFile());
```
Comment (vilchik-elena) [inline sonar-python-plugin/src/test/java/org/sonar/plugins/python/PythonHighlighterTest.java]: @ivandalbosco we could extract path `src/test/resources/org/sonar/plugins/python` used twice in this file

Comment (vilchik-elena) [inline sonar-python-plugin/src/test/java/org/sonar/plugins/python/PythonHighlighterTest.java]: @ivandalbosco I think we can remove `.getAbsoluteFile()`

File: sonar-python-plugin/src/test/java/org/sonar/plugins/python/PythonHighlighterTest.java [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+    
+    assertThat(typeOfTexts.get(0)).isEqualTo(TypeOfText.COMMENT);
+  }
+  
+  private List<TypeOfText>  getTypesOfText(int line, int column, int expectedSize) {
+    List<TypeOfText> typeOfTexts = context.highlightingTypeAt("moduleKey:" + file.getName(), line, column);
+
+    assertThat(typeOfTexts).as("types of text").hasSize(1);
+    
+    return typeOfTexts;
```
Comment (vilchik-elena) [inline sonar-python-plugin/src/test/java/org/sonar/plugins/python/PythonHighlighterTest.java]: @ivandalbosco You better return the first element of this list, then it will be easier to use this method

File: sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonHighlighter.java (hunk 5 above)
Comment (vilchik-elena) [inline sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonHighlighter.java]: @ivandalbosco please be careful with email answers: you commented the wrong line now

File: sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonHighlighter.java [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+ *     </pre>
+ *   </li>
+ *   <li>
+ *     Comments. Example:
+ *     <pre>
+ *        # some comment
+ *     </pre>
+ *   </li>
+ *   <li>
+ *     Doc strings. Examples:
```
Comment (vilchik-elena) [inline sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonHighlighter.java]: @ivandalbosco Do we support this already?

**Comment by vilchik-elena** [comment]
@ivandalbosco btw you don't have to put smth like "incomplete" to PR title. By default if ticket is not "in review" we consider its PR as WIP

File: sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonHighlighter.java (hunk 12 above)
Comment (ivandalbosco) [inline sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonHighlighter.java]: @vilchik-elena OK!

File: sonar-python-plugin/src/test/java/org/sonar/plugins/python/PythonHighlighterTest.java (hunk 13 above)
Comment (ivandalbosco) [inline sonar-python-plugin/src/test/java/org/sonar/plugins/python/PythonHighlighterTest.java]: @vilchik-elena yep, it works

File: sonar-python-plugin/src/test/resources/org/sonar/plugins/python/highlighting.py (hunk 8 above)
Comment (ivandalbosco) [inline sonar-python-plugin/src/test/resources/org/sonar/plugins/python/highlighting.py]: Wrong Java habit...

File: sonar-python-plugin/src/test/java/org/sonar/plugins/python/PythonHighlighterTest.java (hunk 9 above)
Comment (ivandalbosco) [inline sonar-python-plugin/src/test/java/org/sonar/plugins/python/PythonHighlighterTest.java]: @vilchik-elena OK

File: sonar-python-plugin/src/test/resources/org/sonar/plugins/python/highlighting.py (hunk 8 above)
Comment (vilchik-elena) [inline sonar-python-plugin/src/test/resources/org/sonar/plugins/python/highlighting.py]: @ivandalbosco You can name it "trailing"

File: sonar-python-plugin/src/test/resources/org/sonar/plugins/python/highlighting.py [hunk 16]
```diff
@@ -0,0 +1,12 @@
+def foo():
+    pass
+
+x = "string"
```
Comment (vilchik-elena) [inline sonar-python-plugin/src/test/resources/org/sonar/plugins/python/highlighting.py]: @ivandalbosco assert all kinds of string literal

File: sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonHighlighter.java [hunk 17]
```diff
@@ -0,0 +1,121 @@
...
+
+  @Override
+  public void visitToken(Token token) {
+    if (token.getType().equals(PythonTokenType.STRING)) {
+      // case: string literal
+      highlight(token, TypeOfText.STRING);
+    } else if (token.getType() instanceof PythonKeyword) {
+      // case: keyword
+      highlight(token, TypeOfText.KEYWORD);
+    } else if (token.getType().equals(GenericTokenType.COMMENT)) {
```
Comment (vilchik-elena) [inline sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonHighlighter.java]: @ivandalbosco drop this dead code

File: sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonHighlighter.java (hunk 15 above)
Comment (vilchik-elena) [inline sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonHighlighter.java]: @ivandalbosco Please add "they are highlighted the same way as simple string litetals"

File: sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonHighlighter.java [hunk 18]
```diff
@@ -0,0 +1,121 @@
...
+import org.sonar.python.api.PythonTokenType;
+import org.sonar.squidbridge.SquidAstVisitor;
+
+/**
+ * Colorizes Python code. Currently colorizes:
+ * <ul>
+ *   <li>
+ *     String literals. Example:
+ *     <pre>
+ *       "hello"
```
Comment (vilchik-elena) [inline sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonHighlighter.java]: @ivandalbosco add `'foo'`

File: sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonHighlighter.java [hunk 19]
```diff
@@ -0,0 +1,121 @@
...
+ *   </li>
+ *   <li>
+ *     Doc strings. Examples:
+ *     <pre>
+ *        """ a doc string"""
+ *        ''' another doc string '''
+ *     </pre>
+ *   </li>
+ * </ul>
+ * Note that doc strings, such as:
```
Comment (vilchik-elena) [inline sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonHighlighter.java]: @ivandalbosco drop this block

File: sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonHighlighter.java [hunk 20]
```diff
@@ -0,0 +1,121 @@
...
+ * are not supported (i.e., not colorized) yet.
+ */
+public class PythonHighlighter extends SquidAstVisitor<Grammar> implements AstAndTokenVisitor {
+
+  private NewHighlighting newHighlighting;
+  
+  private final SensorContext context;
+  
+  /**
+   * Creates a highlighter for the specified context.
```
Comment (vilchik-elena) [inline sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonHighlighter.java]: @ivandalbosco we can drop this

File: sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonHighlighter.java [hunk 21]
```diff
@@ -0,0 +1,121 @@
...
+  @Override
+  public void visitFile(@Nullable AstNode astNode) {
+    newHighlighting = context.newHighlighting();
+    InputFile inputFile = context.fileSystem().inputFile(context.fileSystem().predicates().is(getContext().getFile().getAbsoluteFile()));
+    newHighlighting.onFile(inputFile);
+  }
+
+  @Override
+  public void visitToken(Token token) {
+    if (token.getType().equals(PythonTokenType.STRING)) {
```
Comment (vilchik-elena) [inline sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonHighlighter.java]: @ivandalbosco please add empty line before `else if` for readability

File: sonar-python-plugin/src/test/java/org/sonar/plugins/python/PythonHighlighterTest.java [hunk 22]
```diff
@@ -0,0 +1,127 @@
...
+  private SensorContextTester context;
+  
+  private File file; 
+  
+  @Before
+  @SuppressWarnings("unchecked")
+  public void scanFile() {
+    String dir = "src/test/resources/org/sonar/plugins/python";
+    
+    file = new File(dir + "/highlighting.py");
```
Comment (vilchik-elena) [inline sonar-python-plugin/src/test/java/org/sonar/plugins/python/PythonHighlighterTest.java]: @ivandalbosco you can use another constructor of `File`: `new File(dir, "highlighting.py")`. It's less error-prone

File: sonar-python-plugin/src/test/java/org/sonar/plugins/python/PythonHighlighterTest.java [hunk 23]
```diff
@@ -0,0 +1,127 @@
...
+    
+    context = SensorContextTester.create(new File(dir));
+    context.fileSystem().add(inputFile);
+    
+    PythonHighlighter pythonHighlighter = new PythonHighlighter(context);
+    PythonAstScanner.scanSingleFile(file, pythonHighlighter);
+  }
+
+  @Test
+  public void keyword_def() throws Exception {
```
Comment (vilchik-elena) [inline sonar-python-plugin/src/test/java/org/sonar/plugins/python/PythonHighlighterTest.java]: @ivandalbosco imho it's better to have 1 test method per highlighted category, thus `keywords` (which merges this and 2 next test methods), `string_literal` and `comments`.

File: sonar-python-plugin/src/test/java/org/sonar/plugins/python/PythonHighlighterTest.java (hunk 22 above)
Comment (ivandalbosco) [inline sonar-python-plugin/src/test/java/org/sonar/plugins/python/PythonHighlighterTest.java]: @vilchik-elena yes we can, but let's not go to much into details...

File: sonar-python-plugin/src/test/resources/org/sonar/plugins/python/highlighting.py [hunk 24]
```diff
@@ -0,0 +1,16 @@
...
+
+# This is a comment
+
+def foo():
+    pass  # other comment
+
+# the line below has 4 keywords: if, or, or, continue
+if t0 == t1 or t1 == t2 or t2 == t0: continue
+
+''' comment '''
```
Comment (vilchik-elena) [inline sonar-python-plugin/src/test/resources/org/sonar/plugins/python/highlighting.py]: @ivandalbosco why `comment` is content of this string?

File: sonar-python-plugin/src/test/resources/org/sonar/plugins/python/highlighting.py (hunk 16 above)
Comment (vilchik-elena) [inline sonar-python-plugin/src/test/resources/org/sonar/plugins/python/highlighting.py]: @ivandalbosco still no `'string'`

File: sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonSquidSensor.java [hunk 25]
```diff
@@ -97,6 +97,7 @@ public void execute(SensorContext context) {
 
     List<SquidAstVisitor<Grammar>> visitors = Lists.newArrayList(checks.all());
     visitors.add(new FileLinesVisitor(fileLinesContextFactory, context.fileSystem(), linesOfCode));
+    visitors.add(new PythonHighlighter(context));
```
Comment (vilchik-elena) [inline sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonSquidSensor.java]: @ivandalbosco i think it would be nice to have at least one assertion about highlighting in PythonSquidSensorTest

Comment (ivandalbosco) [inline sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonSquidSensor.java]: @vilchik-elena should we check that the scanner has a PythonHighlighter visitor?

File: its/ruling/src/test/java/org/sonar/python/it/PythonRulingTest.java [hunk 26]
```diff
@@ -59,7 +59,7 @@ public void test() throws Exception {
       .setProperty("dump.old", FileLocation.of("src/test/resources/expected").getFile().getAbsolutePath())
       .setProperty("dump.new", FileLocation.of("target/actual").getFile().getAbsolutePath())
       .setProperty("sonar.cpd.skip", "true")
-      .setProperty("sonar.analysis.mode", "preview")
+ //     .setProperty("sonar.analysis.mode", "preview")
```
Comment (vilchik-elena) [inline its/ruling/src/test/java/org/sonar/python/it/PythonRulingTest.java]: @ivandalbosco we should undo this change

Comment (ivandalbosco) [inline its/ruling/src/test/java/org/sonar/python/it/PythonRulingTest.java]: Sure

File: sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonHighlighter.java [hunk 27]
```diff
@@ -0,0 +1,137 @@
...
+  @Override
+  public void leaveFile(@Nullable AstNode astNode) {
+    newHighlighting.save();
+  }
+
+  private void highlight(Token token, TypeOfText typeOfText) {
+    TokenLocation tokenLocation = new TokenLocation(token);
+    newHighlighting.highlight(tokenLocation.startLine(), tokenLocation.startLineOffset(), tokenLocation.endLine(), tokenLocation.endLineOffset(), typeOfText);
+
+    // if (typeOfText == TypeOfText.STRING) {
```
Comment (SonarTech) [inline sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonHighlighter.java]: MAJOR This block of commented-out lines of code should be removed. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

File: sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonHighlighter.java [hunk 28]
```diff
@@ -0,0 +1,137 @@
...
+    // firstColumn = token.getColumn();
+    // } else {
+    // line++;
+    // firstColumn = 0;
+    // }
+    // newHighlighting.highlight(line, firstColumn, line, firstColumn + tokenLines[i].length(), typeOfText);
+    // }
+    // } else {
+    // newHighlighting.highlight(token.getLine(), token.getColumn(), token.getLine(), token.getColumn() + token.getValue().length(),
+    // typeOfText);
```
Comment (SonarTech) [inline sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonHighlighter.java]: MAJOR This block of commented-out lines of code should be removed. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* MAJOR 2 major

Watch the comments in this conversation to review them.
